tests/bsim/bluetooth/host/central/src/main.c

```diff
@@ -82,13 +100,116 @@ static void test_central_connect_timeout(void)
 	TEST_PASS("Correct timeout");
 }
 
+static void test_central_connect_when_connecting(void)
+{
+	int err;
+
+	bt_conn_cb_register(&conn_cb);
+
+	/* Initialize Bluetooth */
+	err = bt_enable(NULL);
+	TEST_ASSERT(err == 0, "Can't enable Bluetooth (err %d)", err);
+
+	struct bt_conn *conn;
+
+	bt_addr_le_t peer = {.a.val = {0x01}};
+
+	const struct bt_conn_le_create_param create_param = {
+		.options = BT_CONN_LE_OPT_NONE,
+		.interval = BT_GAP_SCAN_FAST_INTERVAL,
+		.window = BT_GAP_SCAN_FAST_WINDOW,
+	};
+
+	k_sem_reset(&sem_connected);
+
+	err = bt_conn_le_create(&peer, &create_param, BT_LE_CONN_PARAM_DEFAULT, &conn);
+	TEST_ASSERT(err == 0, "Failed starting initiator (err %d)", err);
+
+	/* Now we have a valid connection reference */
+	atomic_val_t initial_refs = atomic_get(&conn->ref);
+
+	TEST_ASSERT(initial_refs >= 1, "Expect to have at least once reference");
+
+	err = bt_conn_le_create(&peer, &create_param, BT_LE_CONN_PARAM_DEFAULT, &conn);
+	TEST_ASSERT(err == -EALREADY, "Expected to fail to create connection (err %d)", err);
+
+	/* Expect the number of refs to be unchanged. */
+	TEST_ASSERT(atomic_get(&conn->ref) == initial_refs,
+			       "Expect number of references to be unchanged");
+
+	err = k_sem_take(&sem_connected, K_FOREVER);
+	TEST_ASSERT(err == 0, "Failed getting connected timeout", err);
+
+	TEST_ASSERT(atomic_get(&conn->ref) == 0, "Expect no more references");
+
+	TEST_PASS("Passed");
+}
+
+static void test_central_connect_to_existing(void)
+{
+	int err;
+
+	bt_conn_cb_register(&conn_cb);
+
+	/* Initialize Bluetooth */
+	err = bt_enable(NULL);
+	TEST_ASSERT(err == 0, "Can't enable Bluetooth (err %d)", err);
+
+	struct bt_conn *conn;
+
+	bt_addr_le_t peer = {.type = BT_ADDR_LE_RANDOM,
+			     .a.val = {0xc0, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0}};
+
+	const struct bt_conn_le_create_param create_param = {
+		.options = BT_CONN_LE_OPT_NONE,
+		.interval = BT_GAP_SCAN_FAST_INTERVAL,
+		.window = BT_GAP_SCAN_FAST_WINDOW,
+	};
+
+	k_sem_reset(&sem_connected);
+
+	err = bt_conn_le_create(&peer, &create_param, BT_LE_CONN_PARAM_DEFAULT, &conn);
+	TEST_ASSERT(err == 0, "Failed starting initiator (err %d)", err);
+
+	err = k_sem_take(&sem_connected, K_FOREVER);
+	TEST_ASSERT(err == 0, "Failed establishing connection", err);
+
+	/* Now we have a valid connection reference */
+	atomic_val_t initial_refs = atomic_get(&conn->ref);
+
+	TEST_ASSERT(initial_refs >= 1, "Expect to have at least once reference");
+
+	err = bt_conn_le_create(&peer, &create_param, BT_LE_CONN_PARAM_DEFAULT, &conn);
```

Comment:
Here the API doesn't protect the user from providing a connection object that still has references to it. I don't think this can be handled with todays implementation unless we change the API to require `conn` to point to NULL.

Reply:
I agree. This is why testlib (which btw is made to work on target) does this. And I now think it's worth the API break to start enforcing this in core API.

https://github.com/zephyrproject-rtos/zephyr/blob/884a4e5a35f3aabd804f74b9e6f31a25eda5be7c/tests/bluetooth/common/testlib/include/testlib/conn.h#L30-L31

Reply:
I don't think it requires an API change. 

Isn't to possible to just do something along the lines of 
```c
if (ret_conn != NULL && atomic_get((*ret_conn)->ref) != 0) {
    return -EALREADY;
}
```

Reply:
isn't that `-EALREADY` an API change though?

Reply:
> isn't that `-EALREADY` an API change though?

Yeah, but not a breaking API change IMO, although that is clearly debatable :)

Reply:
> I don't think it requires an API change.
> 
> Isn't to possible to just do something along the lines of
> 
> ```c
> if (ret_conn != NULL && atomic_get((*ret_conn)->ref) != 0) {
>     return -EALREADY;
> }
> ```

I'm wondering if we really can do this. Consider the following code snippet:
```c
int err;
struct bt_conn conn;

err = bt_conn_le_create(..., &conn);
```
Here is _can_ happen than `conn` is initialized to an existing connection. That is, we cannot check the content of the connection object

Reply:
right, it's a possible busfault on target

Reply:
> ```c
> int err;
> struct bt_conn conn;
> 
> err = bt_conn_le_create(..., &conn);
> ```
> 
> Here is _can_ happen than `conn` is initialized to an existing connection. That is, we cannot check the content of the connection object

The application cannot write this I think since `struct bt_conn` is only exposed in the public api as an opaque pointer. The application is not allowed to instantiate a `struct bt_conn` object.

Reply:
Furthermore, we can check if the pointer is in the `acl_conns` array by using `IS_ARRAY_ELEMENT`.

We should always be able to determine if the pointer provided by the caller is a proper `bt_conn *` or not. If it is, we can check the ref-counter. If it is not, then we either reject if non-NULL or just use it.

Reply:
When declaring `struct bt_conn *conn;`, it can happen that it points to something inside `acl_conns`, right?

Reply:
@hermabe , you are right. The code sample was wrong. I updated the code to declare `conn` as a pointer.

Reply:
Your point still stands with the updated sample. `struct bt_conn *conn;` may point to garbage so `**ret_conn` cannot be used.

Reply:
> When declaring `struct bt_conn *conn;`, it can happen that it points to something inside `acl_conns`, right?

Technically, yes if it's uninitialized. `IS_ARRAY_ELEMENT` should also check whether it's aligned properly, so even if it's within the array, if it's not aligned properly it should still work. 

If it happens to actually match a connection object pointer directly, then it should also be OK as we can then treat it as a proper `bt_conn` pointer and check the reference counter

Reply:
> If it happens to actually match a connection object pointer directly, then it should also be OK as we can then treat it as a proper bt_conn pointer and check the reference counter

I would argue that we shouldn't rely on the undefined initialization value of a pointer.

Lets say the application is configured to support two connections.

```
int err;
struct bt_conn *conn1 = NULL;
err = bt_conn_le_create(..., &conn1);

/* ... establish connection 1 */

int err;
struct bt_conn *conn2; /* By chance this is initialized to conn1 */
err = bt_conn_le_create(..., &conn2); /* This call will now fail */ 
```

I assume that is not intended.

Reply:
> > If it happens to actually match a connection object pointer directly, then it should also be OK as we can then treat it as a proper bt_conn pointer and check the reference counter
> 
> I would argue that we shouldn't rely on the undefined initialization value of a pointer.
> 
> Lets say the application is configured to support two connections.
> 
> ```
> int err;
> struct bt_conn *conn1 = NULL;
> err = bt_conn_le_create(..., &conn1);
> 
> /* ... establish connection 1 */
> 
> int err;
> struct bt_conn *conn2; /* By chance this is initialized to conn1 */
> err = bt_conn_le_create(..., &conn2); /* This call will now fail */ 
> ```
> 
> I assume that is not intended.

That is a fair point. Ideally it shouldn't be necessary to initialize an `[out]` pointer, and there's no way to check that at runtime I believe. 

The above case would be a very rare instance though.